contributors/design-proposals/containerized-mounter-pod.md

```diff
@@ -0,0 +1,200 @@
+
+To sum it up, it's just a daemon set that spawns privileged pods, running a simple init + a daemon that executes mount utilities as requested by kubelet via gRPC.
+
+## Design
+
+### Configuration of the host OS
+With the [HostPath volume propagation](propagation.md) implemented, we must ensure that `/var/lib/kubelet` is share-able into containers so a pod with mount utilities can mount something there in its container and this mount will be visible to kubelet and docker (or other container engine) on the host.
+We propose:
+*  During startup. kubelet checks that `/var/lib/kubelet` on the host is on a mount with shared mount propagation.
+    *  If not, kubelet makes `/var/lib/kubelet` shareable:
+       ```shell
+       mount --bind /var/lib/kubelet /var/lib/kubelet
+       mount --make-shared /var/lib/kubelet
+       ```
+       Note that these commands are executed on the host in case kubelet runs in a container.
+       If this startup adjustment fails, kubelet refuses to start with a clear error message.
+* Kubelet does not do any other special check. We assume that kubelet either runs on the host and in the same mount namespace as docker and the host, or it runs in a container that sees /var/lib/kubelet as it is on the host and with shared mount propagation.
+
+This ensures that kubelet runs out of the box on any distro without any configuration done by the cluster admin.
+
+### Volume plugins
+* All volume plugins need to be updated to use a new `mount.Exec` interface to call external utilities like `mount`, `mkfs`, `rbd lock` and such. Implementation of the interface will be provided by caller and will lead either to simple `os.exec` on the host or a gRPC call to a socket in `/var/lib/kubelet/plugin-sockets/` directory.
+
+### Controllers
+TODO: how will controller-manager talk to a remote pod? It's relatively easy to do something like `kubectl exec <mount pod>` from controller-manager, however it's harder to *discover* the right pod.
+
+### Kubelet
+* When kubelet talks to a volume plugin, it looks for a socket named `/var/lib/kubelet/plugin-sockets/<plugin-name>`. This allows for easier discovery of flex volume drivers - probe in https://github.com/kubernetes/community/pull/833 needs to scan `/var/lib/kubelet/plugin-sockets/` too and find sockets in any new subdirectories.
+  * If the socket does not exist, kubelet gives the volume plugin plain `os.Exec` as implementation of `mount.Exec` interface and all mount utilities are executed on the host.
+  * If the socket exists, kubelet gives the volume plugin `GRPCExec` as implementation of `mount.Exec` and all mount utilities are executed via gRPC on the socket which presumably leads to a pod with mount utilities running a gRPC server.
+
+As consequence, kubelet will try to run mount utilities on the host when it starts and has not received pods with mount utilities yet (and thus `/var/lib/kubelet/plugin-sockets/` is empty). This is likely to fails with a cryptic error:
```

Comment:
I am trying to say that when kubelet starts and it does not run any pods yet (it's a new node), it may receive pods in any order. It can happen that an application pod that requires say GlusterFS volume is placed there *before* our DaemonSet with mount utilities for Gluster is scheduled. 

Our fresh kubelet then looks at /var/lib/kubelet/plugin-sockets, does not see GlusterFS socket there and tries to run mount.glusterfs on the host. This will fail with confusing error message.

Only after a while DaemonSet sees a new node, puts a pod with GlusterFS utilities there and this pod creates a socket in /var/lib/kubelet/plugin-sockets.

Kubelet tries to mount volumes pod periodically, finds the socket there and runs mount.glusterfs in the pod  and everything self-heals. Still, the pod startup is delayed (that's IMO OK) and user can see some errors (which may cause some support calls).

In some old version of this proposal I suggested new kubelet and controller-manager parameter with list of plugins for which kubelet or controller-manager **must** wait for a pod with mount utilities and never try to run them on the host. This was turned down as too complicated.